src/support/modify.c

```diff
@@ -172,6 +178,172 @@ __modify_apply_one(WT_SESSION_IMPL *session, WT_CURSOR *cursor,
+		 * adds branches to the loop applying the changes.
+		 */
+		if (vsize < entries[i].offset + entries[i].size ||
+		    vsize < entries[i].offset + entries[i].data.size)
+			return (0);
+		vsize += (entries[i].data.size - entries[i].size);
+	}
+
+	/* Get a scratch buffer than can hold the result. */
+	WT_RET(__wt_scr_alloc(session, vsize, &tmp));
+
+	/*
+	 * Walk the list of modifications, taking from the original buffer
+	 * anywhere there are gaps. Don't bother checking for copies of 0
+	 * bytes, assume it's unlikely in order to get rid of the branches.
+	 */
+	from = value->data;
+	to = tmp->mem;
+	for (i = 0; i < nentries; ++i) {
+		WT_ASSERT(session,
+		    entries[i].offset >= WT_PTRDIFF(to, tmp->mem));
+		n = entries[i].offset - WT_PTRDIFF(to, tmp->mem);
+		WT_ASSERT(session,
+		    from + n <= (uint8_t *)value->data + value->size);
+
+		memmove(to, from, n);
+		to += n;
+		memmove(to, entries[i].data.data, entries[i].data.size);
+		to += entries[i].data.size;
+
+		from += n + entries[i].size;
+	}
+
+	/* Copy any data remaining in the original buffer. */
+	n = value->size - WT_PTRDIFF(from, value->data);
+	memmove(to, from, n);
+	to += n;
+	tmp->size = WT_PTRDIFF(to, tmp->mem);
+
+	/*
+	 * Copy the result into the original buffer. This function is often
+	 * called using a cursor buffer referencing on-page memory and it's easy
+	 * to overwrite a page. A side-effect of setting the buffer is to ensure
+	 * the buffer's value is in buffer-local memory.
+	 */
+	WT_ERR(__wt_buf_set(session, value, tmp->data, tmp->size));
+
+	/* Restore the trailing nul. */
+	if (sformat) {
+		WT_ERR(__wt_buf_grow(session, value, value->size + 1));
+		((char *)value->data)[value->size++] = '\0';
+	}
+	*successp = true;
+
+err:	__wt_scr_free(session, &tmp);
+	return (ret);
+
+}
+
+/*
+ * __modify_apply --
+ *	Apply a single set of WT_MODIFY changes to a buffer.
+ */
+static int
+__modify_apply(WT_SESSION_IMPL *session,
+    WT_CURSOR *cursor, WT_MODIFY *entries, int nentries)
+{
+	WT_ITEM *value;
+	int i;
+	bool success;
+
+	value = &cursor->value;
+
+	/*
+	 * Grow the buffer first. This function is often called using a cursor
+	 * buffer referencing on-page memory and it's easy to overwrite a page.
+	 * A side-effect of growing the buffer is to ensure the buffer's value
+	 * is in buffer-local memory.
+	 */
+	WT_RET(__wt_buf_grow(session, value, value->size));
+
+	/*
+	 * Try fast path #1.
+	 * Do simple replacement in the existing buffer if the modifications are
+	 * overwriting fixed-sized fields.
+	 *
+	 * If either the data size or replacement count extend past the end of
+	 * the current value, we have to deal with padding bytes. Don't try to
+	 * fast-path padding bytes; it's not common and adds branches to the
+	 * loop applying the changes.
+	 */
+	for (i = 0; i < nentries; ++i) {
+		if (entries[i].data.size != entries[i].size ||
+		    value->size < entries[i].offset + entries[i].size ||
```

Comment:
Since comparison operators are short-circuiting, when the first condition is false, the next two conditions would end up being same, ie entries[i].size equals entries[i].data.size. So we can get rid of one of those?

Reply:
Thanks, fixed in a5fd572.